### PR TITLE
native,testlapack: fix Dlatrs and increase its coverage

### PR DESCRIPTION
**Pull request by vladimir-ch**
Bigger PR than planned but one thing leads to another ...

If you wish I can remove the last commit and submit it separately. I reduced the indentation to make it easier for me to compare with the reference.

It would be nice to export dlattr later but I don't think it's ready.

@kortschak  @btracey  PTAL

### CODE REVIEW

File: testlapack/matgen.go [hunk 1]
```diff
@@ -283,6 +283,31 @@ func Dlagge(m, n, kl, ku int, d []float64, a []float64, lda int, rnd *rand.Rand,
 	// superdiagonals to ku.
 }
 
+// dlarnv fills dst with random numbers from a uniform or normal distribution
+// specified by dist:
+//  dist=1: uniform[0,1),
```
Comment (kortschak) [inline testlapack/matgen.go]: Not that it probably matters that much, but it's not clear from the Fortran docs whether the distributions are e.g. [0,1) or (0,1).

File: testlapack/matgen.go [hunk 2]
```diff
@@ -12,6 +12,15 @@ import (
 	"github.com/gonum/blas/blas64"
 )
 
+const (
```
Comment (kortschak) [inline testlapack/matgen.go]: `dlamchE` is used elsewhere than just `matgen.go`, so these should probably be in `general.go`.

File: native/dlatrs.go [hunk 3]
```diff
@@ -176,16 +176,16 @@ Finish:
 		if noTrans {
 			for j := jfirst; j != jlast; j += jinc {
 				xj := math.Abs(x[j])
-				var tjjs float64
+				var tjj, tjjs float64
```
Comment (kortschak) [inline native/dlatrs.go]: Why change this line?

File: testlapack/matgen.go (hunk 2 above)
Comment (vladimir-ch) [inline testlapack/matgen.go]: I wanted the number to be together, I'll move them to general.go

File: testlapack/matgen.go (hunk 1 above)
Comment (vladimir-ch) [inline testlapack/matgen.go]: Yea, it's not exactly a useful or necessary piece of information. I will change it to parentheses

File: native/dlatrs.go (hunk 3 above)
Comment (vladimir-ch) [inline native/dlatrs.go]: Because the new `goto Skip1` would skip over the declaration of of `tjj`.

**Comment by vladimir-ch** [comment]
> 7989362 confuses me. Can you explain what is going on there?

It actualy fixes two bugs.

One is the for loop bounds. The end condition is `j != jend` so `jend` must be one index past the last column if we want to include `jend`.

The other is that breaking out of the loop was skipping part of the loop body and the remaining iterations. I replaced it with `goto Skip1` .

File: testlapack/matgen.go (hunk 2 above)
Comment (vladimir-ch) [inline testlapack/matgen.go]: Done.

File: testlapack/matgen.go (hunk 1 above)
Comment (vladimir-ch) [inline testlapack/matgen.go]: Done.

File: native/dlatrs.go (hunk 3 above)
Comment (kortschak) [inline native/dlatrs.go]: Ah, things in the spec that get so rarely exercised in my mind.

File: testlapack/general.go [hunk 4]
```diff
@@ -17,8 +17,14 @@ import (
...
-// dlamchE is the machine epsilon. For IEEE this is 2^-53.
-const dlamchE = 1.0 / (1 << 53)
+const (
+	// dlamchE is the machine epsilon. For IEEE this is 2^-53.
+	dlamchE = 1.0 / (1 << 53)
+	dlamchP = 2 * dlamchE
+	dlamchB = 2
+)
+
+var dlamchS = math.Nextafter((4 / math.MaxFloat64), 0)
```
Comment (vladimir-ch) [inline testlapack/general.go]: Would there be agreement to replace this with
```
const dlamchS = 2.22507385850720138309023271733240406421922e-308 // 2^{-1022}
```
? Also in native.

Comment (kortschak) [inline testlapack/general.go]: I'd rather not, though that number can be expressed (horribly) as `1.0 / (1 << 256) / (1 << 256) / (1 << 256) / (1 << 254)`. So maybe use that, with the comment. At least it's easy to add the numbers together and confirm that they are 1022. Then that const can go in the block above.

File: testlapack/matgen.go [hunk 5]
```diff
@@ -308,6 +317,323 @@ func dlarnv(dst []float64, dist int, rnd *rand.Rand) {
 	}
 }
 
+// dlattr generates an n×n triangular test matrix A with its properties uniquely
+// determined by imat and uplo, and returns whether A has unit diagonal. If diag
+// is blas.Unit, the diagonal elements are set so that A[k,k]=k.
+//
+// trans specifies whether the matrix A or its transpose will be used.
```
Comment (kortschak) [inline testlapack/matgen.go]: Note that trans is not currently used or add the final transpose when asked for.

Comment (vladimir-ch) [inline testlapack/matgen.go]: Thanks, I completely forgot about it. I will add it.

File: testlapack/matgen.go [hunk 6]
```diff
@@ -308,6 +317,323 @@ func dlarnv(dst []float64, dist int, rnd *rand.Rand) {
...
+	ulp := dlamchE * dlamchB
+	smlnum := dlamchS
+	bignum := (1 - ulp) / smlnum
+
+	bi := blas64.Implementation()
+
+	switch imat {
+	default:
+		// TODO(vladimir-ch): Implement the remaining cases.
+		panic("testlapack: invalid or not implemented imat")
```
Comment (vladimir-ch) [inline testlapack/matgen.go]: Thanks. Done.

File: testlapack/general.go (hunk 4 above)
Comment (vladimir-ch) [inline testlapack/general.go]: I changed it to const using the expression you suggested. Later we can do the same change in native.

File: testlapack/matgen.go (hunk 5 above)
Comment (vladimir-ch) [inline testlapack/matgen.go]: Done.

**Pull request merged**